File: csharp/Leet/Today1/ValidParentheses.cs
// https://leetcode.com/problems/valid-parentheses/description/

namespace Today1.ValidParentheses;

public class Solution
{
    public bool IsValid(string s)
    {
        var stack = new Stack<char>();
        for (int i = 0; i < s.Length; i++)
        {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[')
                stack.Push(s[i]);
            else
            {
                if (stack.Count == 0)
                    return false;

                var openBraket = stack.Pop();

                if (s[i] == ')')
                {
                    if (openBraket != '(')
                        return false;
                }
                else if (s[i] == '}')
                {
                    if (openBraket != '{')
                        return false;
                }
                else if (s[i] == ']')
                {
                    if (openBraket != '[')
                        return false;
                }
            }
        }
        return stack.Count == 0;
    }
}

File: python/today0/valid_parentheses.py
# https://leetcode.com/problems/valid-parentheses/description/

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mappings = {'}':'{', ')':'(', ']':'['}
        for c in s:
            if c in mappings.values():
                stack.append(c)
            elif not stack:
                return False
            elif stack.pop() != mappings[c]:
                return False
        return not stack
